core/src/main/java/de/danoeh/antennapod/core/export/favorites/FavoritesWriter.java

```diff
@@ -0,0 +1,125 @@
+        String template = IOUtils.toString(templateStream, "UTF-8");
+        template = template.replaceAll("\\{TITLE\\}", "Favorites");
+        String[] templateParts = template.split("\\{FEEDS\\}");
+
+        Map<Long, List<FeedItem>> favoriteByFeed = getFeedMap(getFavorites());
+
+        writer.append(templateParts[0]);
+
+        for (Long feedId : favoriteByFeed.keySet()) {
+            List<FeedItem> favorites = favoriteByFeed.get(feedId);
+            writer.append("<li><div>");
+            writeFeed(writer, favorites.get(0).getFeed());
+
+            writer.append("<ul>");
+            for (FeedItem item : favorites) {
+                writeFavoriteItem(writer, item);
+            }
+            writer.append("</ul></div></li>\n");
+        }
+
+        writer.append(templateParts[1]);
+
+        Log.d(TAG, "Finished writing document");
+    }
+
+    private List<FeedItem> getFavorites() {
+        int page = 0;
+
+        List<FeedItem> favoritesList = new ArrayList<>();
+        List<FeedItem> favoritesPage;
+        do {
+            favoritesPage = DBReader.getFavoriteItemsList(page * PAGE_LIMIT, PAGE_LIMIT);
+            favoritesList.addAll(favoritesPage);
+            ++page;
+        } while (!favoritesPage.isEmpty() && favoritesPage.size() == PAGE_LIMIT);
+
+        // sort in descending order
+        Collections.sort(favoritesList, (lhs, rhs) -> rhs.getPubDate().compareTo(lhs.getPubDate()));
+
+        return favoritesList;
+    }
+
+    /**
+     * Group favorite episodes by feed, sorting them by publishing date in descending order.
+     *
+     * @param favoritesList {@code List} of all favorite episodes.
+     * @return A {@code Map} favorite episodes, keyed by feed ID.
+     */
+    private Map<Long, List<FeedItem>> getFeedMap(List<FeedItem> favoritesList) {
+        Map<Long, List<FeedItem>> feedMap = new TreeMap<>();
+
+        for (FeedItem item : favoritesList) {
+            List<FeedItem> feedEpisodes = feedMap.get(item.getFeedId());
+
+            if (feedEpisodes == null) {
+                feedEpisodes = new ArrayList<>();
+                feedMap.put(item.getFeedId(), feedEpisodes);
+            }
+
+            feedEpisodes.add(item);
+        }
+
+        return feedMap;
+    }
+
+    private void writeFeed(Writer writer, Feed feed) throws IOException {
+        writer.append("<img src=\"");
+        writer.append(feed.getImageUrl());
+        writer.append("\" /><p>");
+        writer.append(feed.getTitle());
+        writer.append(" <span><a href=\"");
+        writer.append(feed.getLink());
+        writer.append("\">Website</a> • <a href=\"");
+        writer.append(feed.getDownload_url());
+        writer.append("\">Feed</a></span></p>");
+    }
+
+    private void writeFavoriteItem(Writer writer, FeedItem item) throws IOException {
+        writer.append("<li><div style=\"border-top: 1px solid #eee8e8; padding: 0px; box-shadow: none\"><span>");
```

Comment:
Could you please remove the span and the div with the inline style and use CSS in the template file instead (I think the `li` should be enough to achieve the same look)? I want to try to mix html and java as little as possible.

Reply:
The `<div>` and `<span>` tags are the only way I could find to get it to display the items left-aligned, and with a separating bar. Without them, it would look like the screenshot below. Apparently, the `<br>` tag is completely ignored as well (**Website** and **Feed** should be on a separate line).

As for separating HTML and Java - maybe we can use template HTML files for the list items as well, and have placeholders for each segment - like what we have in the HTML export template and `{FEEDS}`.

![Screenshot_1589702413](https://user-images.githubusercontent.com/12814657/82139049-dbfd9200-982d-11ea-893c-46819ed36f53.png)

Reply:
> As for separating HTML and Java - maybe we can use template HTML files for the list items as well, and have placeholders for each segment - like what we have in the HTML export template and {FEEDS}.

That sounds pretty nice (if that's not too much work)

> Apparently, the <br> tag is completely ignored as well (Website and Feed should be on a separate line).

Maybe something like `ul ul li {clear: left;}`?

Reply:
I've uploaded a version with feed and favorite templates, as well as separated the CSS code into the main template. I hope this is more like what you hand in mind.

I had to use a bit more elaborate CSS to get it to display properly, but at least it's not in-line now :-)